katsdptelstate/aio/redis.py

```diff
@@ -235,22 +286,37 @@ async def get_range(self, key: bytes, start_time: float, end_time: float,
     async def monitor_keys(self, keys: Iterable[bytes]) -> AsyncGenerator[KeyUpdateBase, None]:
         # Refer to katsdptelstate.redis.RedisBackend for details of the protocol.
         while True:
-            receiver = aioredis.pubsub.Receiver()
-            # We use a dedicated connection for the lifetime of this generator.
-            # While the connection pool also has a member to hold a pubsub
-            # connection, it's more robust to just close a connection than to
-            # try to manage unsubscriptions.
-            client: Optional[aioredis.Redis] = None
+            queue = asyncio.Queue()    # type: asyncio.Queue[_QueueItem]
             try:
-                conn = await self.client.connection.acquire()
-                client = aioredis.Redis(conn)
-                await client.subscribe(*(receiver.channel(b'update/' + key) for key in keys))
+                # We need to serialise subscriptions because otherwise two
+                # monitors that wish to monitor the same key may each create
+                # a channel and try to subscribe to it, and only one of them
+                # will actually receive events. Unsubscriptions do not take
+                # the lock: the worst that can happen is that we get
+                # accidentally unsubscribed after we think we've subscribed,
+                # but that will close the channel and hence we'll try again.
+                async with self._subscribe_lock:
+                    new_channels = []
+                    for key in keys:
+                        channel_name = b'update/' + key
+                        channel = self.client.connection.pubsub_channels.get(channel_name)
+                        if channel is None:
+                            channel = _Channel(channel_name)
+                            new_channels.append(channel)
+                        channel.queues.add(queue)
+                    if new_channels:
+                        await self.client.subscribe(*new_channels)
+                        new_channels = []
```

Comment:
Is it necessary to clear new_channels - don't see how this section can get rerun without it being cleared anyway ? Why not keep it around until the finally and just use this list to drop queues and unsubscribe rather than looping through the keys again ?

Reply:
It was just a premature optimisation to free up memory (since if something goes wrong and the generator doesn't get closed it could hang around for a long time). I'll remove it. I do sometimes wish Python had variables with life-time scoped to the block instead of the function.

For cleanup I want to check channels against what aioredis thinks is subscribed rather than using a saved list, to better deal with corner cases such as another monitor issuing an unsubscribe just as we were subscribing.

Reply:
Fair enough, although I would have expected close_channels to be a subset of new_channels in this case rather than a superset, and thus iterating through new_channels would be more comprehensive rather than less.

Reply:
close_channels need not be a subset of new_channels. Consider the following sequence:
1. Caller A starts monitoring 'foo'. It will subscribe to the channel.
2. Caller B starts monitoring 'foo'. It will add its queue to the channel, but no subscription is necessary.
3. Caller A finishes monitoring. It will remove its queue, but the channel is still being used for B so it will not unsubscribe.
4. Caller B finishes monitoring and unsubscribes.

So in B, `new_channels` is empty but `close_channels` is not.

Reply:
Cool - make sense.